openraft/src/core/admin.rs

```diff
@@ -150,25 +150,47 @@ impl<'a, D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>
         );
     }
 
-    pub fn has_pending_membership_change(&self, members: &BTreeSet<NodeId>) -> bool {
-        if self.core.committed < Some(self.core.effective_membership.log_id) {
-            return false;
-        }
-
-        let curr = &self.core.effective_membership.membership;
-        let all_learners = curr.all_learners();
-        for new_node in members.difference(curr.all_members()) {
-            match all_learners.get(new_node) {
-                Some(_node) => {
-                    continue;
-                }
-                None => {
-                    // if new member not include in the learners, there is pending membership change
-                    return true;
-                }
+    #[tracing::instrument(level = "debug", skip(self))]
+    pub(super) fn make_sure_member_in_learners(&mut self, members: &BTreeSet<NodeId>, curr: Membership) -> Membership {
+        let mut new_learner_set = BTreeSet::new();
+        for node_id in members {
+            let target = *node_id;
+            if target == self.core.id {
+                continue;
+            }
+            if let Some(_t) = self.nodes.get(node_id) {
+                tracing::debug!("target node {} is already a cluster member or is being synced", target);
+                continue;
+            }
+            if curr.contains(node_id) {
+                tracing::debug!(
+                    "target node {} is already a member or learner,cannot add as learner",
+                    target
+                );
+                continue;
             }
+
+            new_learner_set.insert(target);
+
+            let state = self.spawn_replication_stream(target, None);
+            self.nodes.insert(target, state);
         }
-        false
+
+        curr.add_learner_set(new_learner_set)
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    pub(super) fn calc_new_membership(&mut self, members: &BTreeSet<NodeId>) -> Membership {
+        let curr = self.core.effective_membership.membership.clone();
+
+        // make sure all the members in the learners and spawn the replication stream
+        let config = self.make_sure_member_in_learners(members, curr);
```

Comment:
IMHO, it does not require a member to be in `learners`, in our new design.
Just:
- Append a new membership log

- Then spawning/shutdown replication task should be done in `append_membership_log()`: https://github.com/datafuselabs/openraft/blob/b6727c46d4793f2f3623017ec8e9875af66108d8/openraft/src/core/admin.rs#L255-L274
   Because every time membership changes, no matter members or learners, the replication should be updated at onece.

- Then replicate the client request

Reply:
but, in function `change_membership`,if self.nodes.get(new_node_id) fail, the function will return `ChangeMembershipError::LearnerNotFound`, and this check is done before `append_membership_log`.

Reply:
Previously `change_membership()` adds learner manually, because learner is not part of membership.
Now learners become part of membership, the replication creating/removing should be part of membership-change.

Thus let `append_membership_log()` spawns/shutdown replication tasks(more precisely, replication should be updated together with `effective_membership`, instead of before changing `effective_membership`).

Since `append_membership_log()` updates replication to the expected state(e.g., adding or removing `self.nodes`), there is no need to check replication with `if self.nodes.get()...` any more.